source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -137,9 +137,59 @@ void MapdataCheckPoint::clearSearched() {
     m_searched = false;
 }
 
+/// @addr{0x80511EC8}
+/// @brief Finds the offset between the two positions that enter the checkpoint.
+/// @details This assumes the player is entering the checkpoint as intended, and not from the side.
+/// @param prevPos The previous position, likely not located in the checkpoint.
+/// @param pos The current position, likely located in the checkpoint.
+/// @return The earliest subdivision that crosses into the checkpoint, in the range [1, 17].
+u16 MapdataCheckPoint::getEntryOffsetMs(const EGG::Vector2f &prevPos,
+        const EGG::Vector2f &pos) const {
+    constexpr f32 REFRESH_PERIOD = 1000.0f / 59.94f;
+
+    EGG::Vector2f velocity = pos - prevPos;
+    velocity *= 1.0f / REFRESH_PERIOD;
+
+    // d_k = p_0 - m + kv
+    EGG::Vector2f displacement = prevPos - m_midpoint + velocity;
+
+    u16 k = 1;
+    for (; static_cast<f32>(k) < REFRESH_PERIOD; ++k) {
+        if (displacement.dot(m_dir) >= 0.0f) {
+            break;
+        }
+
+        displacement += velocity;
+    }
+
+    return k;
+}
+
+/// @brief Finds the offset between the two positions that enter the checkpoint.
+/// @details This assumes the player is entering the checkpoint as intended, and not from the side.
```

Comment:
Pedantic, but I would also explain that this is not present in the base game, and rather this is used as supplementary information to help people gauge if they are making slight improvements to previous ghosts?